db/db_impl_open.cc

```diff
@@ -479,6 +479,22 @@ Status DBImpl::Recover(
     }
   }
 
+  std::vector<std::string> file_names;
+  if (s.ok()) {
+    s = env_->GetChildren(GetName(), &file_names);
+  }
+  if (s.ok()) {
+    uint64_t number = 0;
+    uint64_t options_file_number = 0;
+    FileType type;
+    for (const auto& fname : file_names) {
+      if (ParseFileName(fname, &number, &type) && type == kOptionsFile) {
+        options_file_number = std::max(number, options_file_number);
+      }
+    }
+    versions_->options_file_number_ = options_file_number;
```

Comment:
Could you explain more on why we need this block of change? I guess I don't understand how this change only impacts checkpoints with a read-only db. Or does this apply to regular (r-w) DBs as well?

Reply:
The reason we want to bring the `options_file_number_` to reflect the most recent OPTIONS file is because it's possible to do a checkpoint on the db immediately after it's opened. Failing to do so can lead to checkpointing thread fail to find the OPTIONS file, thus failing the checkpoint operation. I didn't have this at first, but then I did some test and realized this was necessary.

Reply:
My question was around how this is relevant only to read-only DBs? Shouldn't we be seeing the same problem with rw db as well? 
If its critical enough to be fixed irrespective of ro or rw, then I think we should pull it out into a separate PR so that it gets more visibility.

Reply:
Other than this issue, the PR looks good to me.

Reply:
For regular DB, we will call `DBImpl::Open` -> `WriteOptionsFile` -> `RenameTempFileToOptionsFile` in which we update the `options_file_number_` to be `versions_->NewFileNumber()`. That means, before we can checkpoint on the regular DB, the `options_file_number_` is valid.
In contrast, we call `DB::OpenForReadOnly` to create a readonly db instance. This function does not call `DBImpl::Open`. By the time it returns successfully, `options_file_number_` is still 0. Consequently, trying to checkpoint on the db instance now will complain about the absence of the OPTIONS file because the checkpoint thread tries to copy/link an OPTIONS file named 'OPTIONS-0'.

Reply:
Got it, thanks for that awesome explanation! appreciate it. 
Then, should this block of code be enabled only when the function argument `read_only==true`? Asking since `WriteOptionsFile` will anyway be called in the regular case.